src/System/Hatrace/Format.hs

```diff
@@ -0,0 +1,162 @@
+{-# LANGUAGE OverloadedStrings #-}
+module System.Hatrace.Format
+  ( SyscallEnterFormatting(..)
+  , SyscallExitFormatting(..)
+  , ArgFormatting(..)
+  , FormattedSyscall(..)
+  , FormattedArg(..)
+  , FormattedReturn(..)
+  , StringFormattingOptions(..)
+  , defaultStringFormattingOptions
+  , syscallToString
+  , syscallExitToString
+  , argPlaceholder
+  , formatReturn
+  ) where
+
+import           Data.Aeson
+import           Data.ByteString (ByteString)
+import           Data.List (intercalate)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as TE
+import qualified Data.Text.Encoding.Error as TE
+import           Data.Word (Word64)
+import           Foreign.C.Types (CInt(..), CUInt(..), CLong(..), CULong(..), CSize(..))
+import           System.Posix.Types (CMode(..))
+
+class SyscallEnterFormatting a where
+  syscallEnterToFormatted :: a -> FormattedSyscall
+
+class SyscallExitFormatting a where
+  syscallExitToFormatted :: a -> (FormattedSyscall, FormattedReturn)
+
+data FormattedSyscall =
+  FormattedSyscall SyscallName
+                   [FormattedArg]
+  deriving (Eq, Show)
+
+instance ToJSON FormattedSyscall where
+  toJSON (FormattedSyscall syscallName args) =
+    object [ "name" .= syscallName
+           , "args" .= args
+           ]
+
+type SyscallName = String
+
+class ArgFormatting a where
+  formatArg :: a -> FormattedArg
+
+instance ArgFormatting ByteString where
+  formatArg = VarLengthStringArg . T.unpack . TE.decodeUtf8With TE.lenientDecode
+
+instance ArgFormatting Word64 where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting CInt where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting CUInt where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting CLong where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting CULong where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting CSize where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting CMode where
+  formatArg = IntegerArg . fromIntegral
+
+instance ArgFormatting a => ArgFormatting [a] where
+  formatArg = ListArg . map formatArg
+
+data FormattedArg
+  = IntegerArg Integer -- using Integer to accept both Int64 and Word64 at the same time
+  | FixedStringArg String
+  | VarLengthStringArg String
+  | ListArg [FormattedArg]
+  | StructArg [(StructFieldName, FormattedArg)]
+  deriving (Eq, Show)
+
+instance ToJSON FormattedArg where
+  toJSON arg = case arg of
+    IntegerArg n -> toJSON n
+    FixedStringArg s -> toJSON s
+    VarLengthStringArg s -> toJSON s
+    ListArg xs -> toJSON xs
+    StructArg fieldValues ->
+      object [ T.pack name .= value | (name, value) <- fieldValues ]
+
+data FormattedReturn
+  = NoReturn
+  | FormattedReturn FormattedArg
+  deriving (Eq, Show)
```

Comment:
Let's add `Ord` everywhere as well, so that such things can be put in sets and maps